### PR TITLE
Check domain for auto-ssl in redis before issuing certioficate

### PR DESCRIPTION
**Pull request by coodix**
Hi @parisk !

Could you please have a look at realisation?

I tested same method in my separate repo, it's working. Just want to ensure you are interested in these changes, before investing time to cleaning up code and optimizations.

### CODE REVIEW

File: proxy/nginx/lualib/router.lua [hunk 1]
```diff
@@ -20,7 +19,7 @@ local res, err = red:connect(redis_host, redis_port)
 
 -- Return if could not connect to Redis
 if not res then
-    return
+    return ngx.exit(ngx.HTTP_BAD_GATEWAY)
```
Comment (coodix) [inline proxy/nginx/lualib/router.lua]: These changes are for avoiding errors "[error] 19#19: *2 fallback could not be resolved (3: Host not found)"

File: proxy/nginx/lualib/router.lua [hunk 2]
```diff
@@ -35,7 +34,7 @@ if not res or res == ngx.null then
     key = prefix .. ":routes:$wildcard"
     res, err = red:get(key)
     if not res or res == ngx.null then
-        return
+        return ngx.exit(ngx.HTTP_BAD_GATEWAY)
```
Comment (coodix) [inline proxy/nginx/lualib/router.lua]: These changes are for avoiding errors "[error] 19#19: *2 fallback could not be resolved (3: Host not found)"

**Comment by coodix** [comment]
@parisk , sorry, I expected you just review the approach. Didn't want waste your time. Now it's working.

**Comment by parisk** [comment]
Not a problem at all @coodix. Now, when I try hitting an HTTPS endpoint, I get the following error from cURL:

```
curl: (35) gnutls_handshake() failed: The TLS connection was non-properly terminated.
```

Any idea? Is this the case for you as well?

File: proxy/nginx/lualib/router.lua [hunk 3]
```diff
@@ -1,4 +1,3 @@
-local container_url = ngx.var.container_url
```
Comment (parisk) [inline proxy/nginx/lualib/router.lua]: Could this be causing https://github.com/sourcelair/ceryx/pull/39#issuecomment-358246793?

**Comment by parisk** [comment]
@coodix did you have a chance to take a look at what might be causing https://github.com/sourcelair/ceryx/pull/39#issuecomment-358246793?

**Comment by coodix** [comment]
@parisk , not yet. I'll do this week.

**Comment by parisk** [comment]
Not a problem at all, thanks!

**Comment by coodix** [comment]
@parisk , could please five details how do you test?

**Comment by coodix** [comment]
Is you domain in redis storage?
What curl command do you exectue?
What's in logs of the proxy container?

Can't reproduce same error.

**Comment by parisk** [comment]
OK, I got this jotted down.

1. Get a fresh Ubuntu machine and install the latest:
    - Docker: https://docs.docker.com/install/linux/docker-ce/ubuntu
    - [pip](https://pip.pypa.io/en/stable/): `sudo apt-get -y install python-pip`
    - [Docker Compose](https://docs.docker.com/compose/): `sudo pip install docker-compose`
    - [httpie](https://httpie.org/): `sudo pip install httpie`
2. Clone https://github.com/sourcelair/ceryx
3. Check out the [`temp-allow_domain_by_redis` branch](https://github.com/sourcelair/ceryx/tree/temp-allow_domain_by_redis) *
4. Build the refreshed `sourcelair/ceryx-proxy:xenial` image:
    - `cd proxy`
    - `docker build -t sourcelair/ceryx-proxy:xenial -f Dockerfile.xenial .`
    - `cd ..` (go back)
5. Create an `.env` file with the following line:
    - `CERYX_API_HOSTNAME=put.something.here`
6. Launch ceryx: `docker-compose --build -d`
7. Populate the API: docker-compose exec api bin/populate-api
8. ✅ Try accessing the API using the above hostname via **HTTP** (should work)
9. ⛔️ Try accessing the API using the above hostname via **HTTPS** (should not work)

Here you can also find a screencast of the testing procedure: https://www.dropbox.com/s/oo5i33pfbdfggc5/ceryx-ssl-issue.mov.

\* I created this branch, because we have to perform some additional changes because the `xfs` user got renamed to `www-data` in the latest ubunu:xenial image.

**Comment by coodix** [comment]
@parisk , thanks for such a detailed response! 

What tool do you use as a http client? I can see the commend "http". What is it?

Actually I had similar error when tried to pass Host header and access https://127.0.0.1 via curl. SNI didn't work in that case. I had to pass correct domain as part of url. For sniffing resolution of domains to 127.0.0.1 it's possible to pass --resolve option. Or just put entry to local /etc/hosts.

I'll try to reproduce your case.

**Comment by coodix** [comment]
@parisk , got it. You have wrong volumes in docker-compose.override.yml. They replace configs nginx configs.

**Pull request merged**